LinksProcessor ID field error for native class HalCollectionWrapper

### Describe the bug

Hello,

When I upgrade quarkus from 3.6.9 to 3.7.1 PR @bpasson https://github.com/quarkusio/quarkus/pull/38265/files is causing the following issue.

My entity is :


    import io.quarkus.resteasy.reactive.links.RestLinkId;
    
    @Getter
    @AllArgsConstructor
    @NoArgsConstructor
    public class Car
    {
    
       @RestLinkId
       private int id;
       private int year;
       private String make;
       private String model;
    }


In my controller I have : 

      Collection<Object> objectCollection = returnCars.stream().map(car -> (Object)car).collect(Collectors.toList());
      HalCollectionWrapper halCollection = resteasyReactiveHalService.toHalCollectionWrapper(objectCollection,"result_page_"+pageRequest.getPage()+"_size_"+pageRequest.getSize(),
               Car.class);


When the app starts I get:


        2024-02-01 16:46:50,992 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	        [error]: Build step io.quarkus.resteasy.reactive.links.deployment.LinksProcessor#initializeLinksProvider threw an exception: java.lang.IllegalStateException: Cannot generate web links for the class io.quarkus.hal.HalCollectionWrapper because it is either missing an `id` field, a field with an `@Id` annotation or a field with a `@RestLinkId annotation
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:268)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateClassHasFieldId(LinksProcessor.java:232)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.implementPathParameterValueGetters(LinksProcessor.java:151)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.initializeLinksProvider(LinksProcessor.java:76)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
	        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	        at java.base/java.lang.Thread.run(Thread.java:840)
	        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
        
	        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:334)
	        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:251)
	        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:60)
	        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:112)
	        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:433)
	        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)
	        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
	        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
	        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
	        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
        Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	        [error]: Build step io.quarkus.resteasy.reactive.links.deployment.LinksProcessor#initializeLinksProvider threw an exception: java.lang.IllegalStateException: Cannot generate web links for the class io.quarkus.hal.HalCollectionWrapper because it is either missing an `id` field, a field with an `@Id` annotation or a field with a `@RestLinkId annotation
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:268)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateClassHasFieldId(LinksProcessor.java:232)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.implementPathParameterValueGetters(LinksProcessor.java:151)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.initializeLinksProvider(LinksProcessor.java:76)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
	        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	        at java.base/java.lang.Thread.run(Thread.java:840)
	        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
        
	        at io.quarkus.builder.Execution.run(Execution.java:123)
	        at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:79)
	        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:160)
	        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
	        ... 9 more
        Caused by: java.lang.IllegalStateException: Cannot generate web links for the class io.quarkus.hal.HalCollectionWrapper because it is either missing an `id` field, a field with an `@Id` annotation or a field with a `@RestLinkId annotation
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:268)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateRec(LinksProcessor.java:277)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.validateClassHasFieldId(LinksProcessor.java:232)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.implementPathParameterValueGetters(LinksProcessor.java:151)
	        at io.quarkus.resteasy.reactive.links.deployment.LinksProcessor.initializeLinksProvider(LinksProcessor.java:76)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
	        at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	        at java.base/java.lang.Thread.run(Thread.java:840)
	        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
        
        


### Expected behavior

Since my entity is marked with `RestLinkId` and I am using quarkus's `io.quarkus.hal.HalCollectionWrapper` I should not have received `io.quarkus.resteasy.reactive.links.deployment.LinksProcessor` 's IllegalStateException. HalCollectionWrapper and HalEntityWrapper should be excluded from the Id check.

### Actual behavior

I end up getting 

    java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.resteasy.reactive.links.deployment.LinksProcessor#initializeLinksProvider threw an exception: java.lang.IllegalStateException: Cannot generate web links for the class io.quarkus.hal.HalCollectionWrapper because it is either missing an `id` field, a field with an `@Id` annotation or a field with a `@RestLinkId annotation

### How to Reproduce?

1. Create an entity with RestLinkId annotation
2. Create a controller method like 

     `@GET`
     `@Produces({ MediaType.APPLICATION_JSON, RestMediaType.APPLICATION_HAL_JSON })`
     `@RestLink(rel = "list")`
     `@InjectRestLinks`
       public HalCollectionWrapper getCars(@BeanParam PageRequest pageRequest, @Context UriInfo uriInfo)
       {
          List<Car> returnCars = Cars.cars();
    

         Collection<Object> objectCollection = returnCars.stream().map(car -> (Object)car).collect(Collectors.toList());
      
        HalCollectionWrapper halCollection = resteasyReactiveHalService.toHalCollectionWrapper(objectCollection,"result_page_"+pageRequest.getPage()+"_size_"+pageRequest.getSize(),
         Car.class);

          halCollection.addLinks(Link.fromPath(uriInfo.getAbsolutePath().toString()+"?page="+(pageRequest.getPage()+1)+"size="+pageRequest.getSize()).rel("next_page_url").build());

          halCollection.addLinks(Link.fromPath(uriInfo.getAbsolutePath().toString()+"?page=1&size=20").rel("first_page_url").build());

         halCollection.addLinks(Link.fromPath(uriInfo.getAbsolutePath().toString()).rel("list").build());

         return halCollection;
       }
4. Run application

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

17

### Quarkus version or git rev

3.7

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

8.5

### Additional information

_No response_

--- a/docs/src/main/asciidoc/resteasy-reactive.adoc
+++ b/docs/src/main/asciidoc/resteasy-reactive.adoc
@@ -1869,7 +1869,7 @@ When we call a resource `/records/1` that returns only one instance, then the ou
 }
 ----
 
-Finally, you can also provide additional HAL links programmatically in your resource just by returning either `HalCollectionWrapper` (to return a list of entities) or `HalEntityWrapper` (to return a single object) as described in the following example:
+Finally, you can also provide additional HAL links programmatically in your resource just by returning either `HalCollectionWrapper<T>` (to return a list of entities) or `HalEntityWrapper<T>` (to return a single object) as described in the following example:
 
 [source,java]
 ----
@@ -1877,14 +1877,14 @@ Finally, you can also provide additional HAL links programmatically in your reso
 public class RecordsResource {
 
     @Inject
-    RestLinksProvider linksProvider;
+    HalService halService;
 
     @GET
     @Produces({ MediaType.APPLICATION_JSON, RestMediaType.APPLICATION_HAL_JSON })
     @RestLink(rel = "list")
-    public HalCollectionWrapper getAll() {
+    public HalCollectionWrapper<Record> getAll() {
         List<Record> list = // ...
-        HalCollectionWrapper halCollection = new HalCollectionWrapper(list, "collectionName", linksProvider.getTypeLinks(Record.class));
+        HalCollectionWrapper<Record> halCollection = halService.toHalCollectionWrapper( list, "collectionName", Record.class);
         halCollection.addLinks(Link.fromPath("/records/1").rel("first-record").build());
         return halCollection;
     }
@@ -1894,9 +1894,9 @@ public class RecordsResource {
     @Path("/{id}")
     @RestLink(rel = "self")
     @InjectRestLinks(RestLinkType.INSTANCE)
-    public HalEntityWrapper get(@PathParam("id") int id) {
+    public HalEntityWrapper<Record> get(@PathParam("id") int id) {
         Record entity = // ...
-        HalEntityWrapper halEntity = new HalEntityWrapper(entity, linksProvider.getInstanceLinks(entity));
+        HalEntityWrapper<Record> halEntity = halService.toHalWrapper(entity);
         halEntity.addLinks(Link.fromPath("/records/1/parent").rel("parent-record").build());
         return halEntity;
     }

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapper.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapper.java
@@ -14,25 +14,25 @@ import jakarta.ws.rs.core.Link;
  * - the JSON-B serializer: {@link HalCollectionWrapperJsonbSerializer}
  * - the Jackson serializer: {@link HalCollectionWrapperJacksonSerializer}
  */
-public class HalCollectionWrapper extends HalWrapper {
+public class HalCollectionWrapper<T> extends HalWrapper {
 
-    private final Collection<HalEntityWrapper> collection;
+    private final Collection<HalEntityWrapper<T>> collection;
     private final String collectionName;
 
-    public HalCollectionWrapper(Collection<HalEntityWrapper> collection, String collectionName, Link... links) {
+    public HalCollectionWrapper(Collection<HalEntityWrapper<T>> collection, String collectionName, Link... links) {
         this(collection, collectionName, new HashMap<>());
 
         addLinks(links);
     }
 
-    public HalCollectionWrapper(Collection<HalEntityWrapper> collection, String collectionName, Map<String, HalLink> links) {
+    public HalCollectionWrapper(Collection<HalEntityWrapper<T>> collection, String collectionName, Map<String, HalLink> links) {
         super(links);
 
         this.collection = collection;
         this.collectionName = collectionName;
     }
 
-    public Collection<HalEntityWrapper> getCollection() {
+    public Collection<HalEntityWrapper<T>> getCollection() {
         return collection;
     }
 

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapperJacksonSerializer.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapperJacksonSerializer.java
@@ -6,10 +6,10 @@ import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.databind.JsonSerializer;
 import com.fasterxml.jackson.databind.SerializerProvider;
 
-public class HalCollectionWrapperJacksonSerializer extends JsonSerializer<HalCollectionWrapper> {
+public class HalCollectionWrapperJacksonSerializer extends JsonSerializer<HalCollectionWrapper<?>> {
 
     @Override
-    public void serialize(HalCollectionWrapper wrapper, JsonGenerator generator, SerializerProvider serializers)
+    public void serialize(HalCollectionWrapper<?> wrapper, JsonGenerator generator, SerializerProvider serializers)
             throws IOException {
         generator.writeStartObject();
         writeEmbedded(wrapper, generator, serializers);
@@ -17,7 +17,7 @@ public class HalCollectionWrapperJacksonSerializer extends JsonSerializer<HalCol
         generator.writeEndObject();
     }
 
-    private void writeEmbedded(HalCollectionWrapper wrapper, JsonGenerator generator, SerializerProvider serializers)
+    private void writeEmbedded(HalCollectionWrapper<?> wrapper, JsonGenerator generator, SerializerProvider serializers)
             throws IOException {
         JsonSerializer<Object> entitySerializer = serializers.findValueSerializer(HalEntityWrapper.class);
 
@@ -25,14 +25,14 @@ public class HalCollectionWrapperJacksonSerializer extends JsonSerializer<HalCol
         generator.writeStartObject();
         generator.writeFieldName(wrapper.getCollectionName());
         generator.writeStartArray();
-        for (HalEntityWrapper entity : wrapper.getCollection()) {
+        for (HalEntityWrapper<?> entity : wrapper.getCollection()) {
             entitySerializer.serialize(entity, generator, serializers);
         }
         generator.writeEndArray();
         generator.writeEndObject();
     }
 
-    private void writeLinks(HalCollectionWrapper wrapper, JsonGenerator generator) throws IOException {
+    private void writeLinks(HalCollectionWrapper<?> wrapper, JsonGenerator generator) throws IOException {
         generator.writeFieldName("_links");
         generator.writeObject(wrapper.getLinks());
     }

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapperJsonbSerializer.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalCollectionWrapperJsonbSerializer.java
@@ -4,6 +4,8 @@ import jakarta.json.bind.serializer.JsonbSerializer;
 import jakarta.json.bind.serializer.SerializationContext;
 import jakarta.json.stream.JsonGenerator;
 
+// Using the raw type here as eclipse yasson doesn't like custom serializers for
+// generic root types, see https://github.com/eclipse-ee4j/yasson/issues/639
 public class HalCollectionWrapperJsonbSerializer implements JsonbSerializer<HalCollectionWrapper> {
 
     @Override
@@ -14,19 +16,19 @@ public class HalCollectionWrapperJsonbSerializer implements JsonbSerializer<HalC
         generator.writeEnd();
     }
 
-    private void writeEmbedded(HalCollectionWrapper wrapper, JsonGenerator generator, SerializationContext context) {
+    private void writeEmbedded(HalCollectionWrapper<?> wrapper, JsonGenerator generator, SerializationContext context) {
         generator.writeKey("_embedded");
         generator.writeStartObject();
         generator.writeKey(wrapper.getCollectionName());
         generator.writeStartArray();
-        for (HalEntityWrapper entity : wrapper.getCollection()) {
+        for (HalEntityWrapper<?> entity : wrapper.getCollection()) {
             context.serialize(entity, generator);
         }
         generator.writeEnd();
         generator.writeEnd();
     }
 
-    private void writeLinks(HalCollectionWrapper wrapper, JsonGenerator generator, SerializationContext context) {
+    private void writeLinks(HalCollectionWrapper<?> wrapper, JsonGenerator generator, SerializationContext context) {
         context.serialize("_links", wrapper.getLinks(), generator);
     }
 }

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapper.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapper.java
@@ -12,23 +12,23 @@ import jakarta.ws.rs.core.Link;
  * - the JSON-B serializer: {@link HalEntityWrapperJsonbSerializer}
  * - the Jackson serializer: {@link HalEntityWrapperJacksonSerializer}
  */
-public class HalEntityWrapper extends HalWrapper {
+public class HalEntityWrapper<T> extends HalWrapper {
 
-    private final Object entity;
+    private final T entity;
 
-    public HalEntityWrapper(Object entity, Link... links) {
+    public HalEntityWrapper(T entity, Link... links) {
         this(entity, new HashMap<>());
 
         addLinks(links);
     }
 
-    public HalEntityWrapper(Object entity, Map<String, HalLink> links) {
+    public HalEntityWrapper(T entity, Map<String, HalLink> links) {
         super(links);
 
         this.entity = entity;
     }
 
-    public Object getEntity() {
+    public T getEntity() {
         return entity;
     }
 }

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapperJacksonSerializer.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapperJacksonSerializer.java
@@ -12,7 +12,7 @@ import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
 import com.fasterxml.jackson.databind.introspect.BasicClassIntrospector;
 import com.fasterxml.jackson.databind.introspect.BeanPropertyDefinition;
 
-public class HalEntityWrapperJacksonSerializer extends JsonSerializer<HalEntityWrapper> {
+public class HalEntityWrapperJacksonSerializer extends JsonSerializer<HalEntityWrapper<?>> {
 
     @Override
     public void serialize(HalEntityWrapper wrapper, JsonGenerator generator, SerializerProvider serializers)

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapperJsonbSerializer.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalEntityWrapperJsonbSerializer.java
@@ -10,6 +10,8 @@ import org.eclipse.yasson.internal.ProcessingContext;
 import org.eclipse.yasson.internal.model.ClassModel;
 import org.eclipse.yasson.internal.model.PropertyModel;
 
+// Using the raw type here as eclipse yasson doesn't like custom serializers for
+// generic root types, see https://github.com/eclipse-ee4j/yasson/issues/639
 public class HalEntityWrapperJsonbSerializer implements JsonbSerializer<HalEntityWrapper> {
 
     @Override

--- a/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalService.java
+++ b/extensions/hal/runtime/src/main/java/io/quarkus/hal/HalService.java
@@ -24,10 +24,10 @@ public abstract class HalService {
      * @param entityClass The class of the objects in the collection. If null, it will not resolve the links for these objects.
      * @return The Hal collection wrapper instance.
      */
-    public HalCollectionWrapper toHalCollectionWrapper(Collection<Object> collection, String collectionName,
+    public <T> HalCollectionWrapper<T> toHalCollectionWrapper(Collection<T> collection, String collectionName,
             Class<?> entityClass) {
-        List<HalEntityWrapper> items = new ArrayList<>();
-        for (Object entity : collection) {
+        List<HalEntityWrapper<T>> items = new ArrayList<>();
+        for (T entity : collection) {
             items.add(toHalWrapper(entity));
         }
 
@@ -36,7 +36,7 @@ public abstract class HalService {
             classLinks = getClassLinks(entityClass);
         }
 
-        return new HalCollectionWrapper(items, collectionName, classLinks);
+        return new HalCollectionWrapper<>(items, collectionName, classLinks);
     }
 
     /**
@@ -45,8 +45,8 @@ public abstract class HalService {
      * @param entity The entity to wrap.
      * @return The Hal entity wrapper.
      */
-    public HalEntityWrapper toHalWrapper(Object entity) {
-        return new HalEntityWrapper(entity, getInstanceLinks(entity));
+    public <T> HalEntityWrapper<T> toHalWrapper(T entity) {
+        return new HalEntityWrapper<>(entity, getInstanceLinks(entity));
     }
 
     /**

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/pom.xml
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/pom.xml
@@ -23,6 +23,11 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hal-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5-internal</artifactId>
             <scope>test</scope>
         </dependency>

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/HalWrapperResource.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/HalWrapperResource.java
@@ -1,0 +1,63 @@
+package io.quarkus.resteasy.reactive.links.deployment;
+
+import java.util.List;
+
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.PathParam;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.Link;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.UriInfo;
+
+import org.jboss.resteasy.reactive.common.util.RestMediaType;
+
+import io.quarkus.hal.HalCollectionWrapper;
+import io.quarkus.hal.HalEntityWrapper;
+import io.quarkus.hal.HalService;
+import io.quarkus.resteasy.reactive.links.InjectRestLinks;
+import io.quarkus.resteasy.reactive.links.RestLink;
+import io.quarkus.resteasy.reactive.links.RestLinkType;
+
+@Path("/hal")
+public class HalWrapperResource {
+
+    @Inject
+    HalService halService;
+
+    @GET
+    @Produces({ MediaType.APPLICATION_JSON, RestMediaType.APPLICATION_HAL_JSON })
+    @RestLink(rel = "list")
+    @InjectRestLinks
+    public HalCollectionWrapper<TestRecordWithIdAndPersistenceIdAndRestLinkId> getRecords(@Context UriInfo uriInfo) {
+        List<TestRecordWithIdAndPersistenceIdAndRestLinkId> items = List.of(
+                new TestRecordWithIdAndPersistenceIdAndRestLinkId(1, 10, 100, "one"),
+                new TestRecordWithIdAndPersistenceIdAndRestLinkId(2, 20, 200, "two"));
+
+        HalCollectionWrapper<TestRecordWithIdAndPersistenceIdAndRestLinkId> halCollection = halService.toHalCollectionWrapper(
+                items,
+                "collectionName", TestRecordWithIdAndPersistenceIdAndRestLinkId.class);
+        halCollection.addLinks(
+                Link.fromUriBuilder(uriInfo.getBaseUriBuilder().path(String.format("/hal/%d", 1))).rel("first-record").build());
+
+        return halCollection;
+    }
+
+    @GET
+    @Path("/{id}")
+    @Produces({ MediaType.APPLICATION_JSON, RestMediaType.APPLICATION_HAL_JSON })
+    @RestLink(rel = "self")
+    @InjectRestLinks(RestLinkType.INSTANCE)
+    public HalEntityWrapper<TestRecordWithIdAndPersistenceIdAndRestLinkId> getRecord(@PathParam("id") int id,
+            @Context UriInfo uriInfo) {
+
+        HalEntityWrapper<TestRecordWithIdAndPersistenceIdAndRestLinkId> halEntity = halService.toHalWrapper(
+                new TestRecordWithIdAndPersistenceIdAndRestLinkId(1, 10, 100, "one"));
+        halEntity.addLinks(Link.fromUriBuilder(uriInfo.getBaseUriBuilder().path(String.format("/hal/%d/parent", id)))
+                .rel("parent-record").build());
+
+        return halEntity;
+    }
+}

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/HalWrapperResourceTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive-links/deployment/src/test/java/io/quarkus/resteasy/reactive/links/deployment/HalWrapperResourceTest.java
@@ -1,0 +1,65 @@
+package io.quarkus.resteasy.reactive.links.deployment;
+
+import static io.restassured.RestAssured.given;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import jakarta.ws.rs.core.HttpHeaders;
+
+import org.jboss.resteasy.reactive.common.util.RestMediaType;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.builder.Version;
+import io.quarkus.maven.dependency.Dependency;
+import io.quarkus.test.QuarkusUnitTest;
+import io.quarkus.test.common.http.TestHTTPResource;
+import io.restassured.response.Response;
+
+public class HalWrapperResourceTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest TEST = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(HalWrapperResource.class, TestRecordWithIdAndPersistenceIdAndRestLinkId.class))
+            .setForcedDependencies(List.of(
+                    Dependency.of("io.quarkus", "quarkus-resteasy-reactive-jackson", Version.getVersion()),
+                    Dependency.of("io.quarkus", "quarkus-hal", Version.getVersion())));
+
+    @TestHTTPResource("hal")
+    String recordsUrl;
+
+    @TestHTTPResource("hal/{id}")
+    String recordIdUrl;
+
+    @Test
+    void shouldGetAllRecordsWithCustomHalMetadata() {
+        Response response = given()
+                .header(HttpHeaders.ACCEPT, RestMediaType.APPLICATION_HAL_JSON)
+                .get(recordsUrl).thenReturn();
+
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][0].restLinkId")).isEqualTo("1");
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][0]._links.self.href")).endsWith("/hal/1");
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][0]._links.list.href")).endsWith("/hal");
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][1].restLinkId")).isEqualTo("2");
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][1]._links.self.href")).endsWith("/hal/2");
+        assertThat(response.body().jsonPath().getString("_embedded['collectionName'][1]._links.list.href")).endsWith("/hal");
+        assertThat(response.body().jsonPath().getString("_links.first-record.href")).endsWith("/hal/1");
+        assertThat(response.body().jsonPath().getString("_links.list.href")).endsWith("/hal");
+    }
+
+    @Test
+    void shouldGetSingleRecordWithCustomHalMetadata() {
+        Response response = given()
+                .header(HttpHeaders.ACCEPT, RestMediaType.APPLICATION_HAL_JSON)
+                .get(recordIdUrl, 1L)
+                .thenReturn();
+
+        assertThat(response.body().jsonPath().getString("restLinkId")).isEqualTo("1");
+        assertThat(response.body().jsonPath().getString("_links.parent-record.href")).endsWith("/hal/1/parent");
+        assertThat(response.body().jsonPath().getString("_links.self.href")).endsWith("/hal/1");
+        assertThat(response.body().jsonPath().getString("_links.list.href")).endsWith("/hal");
+
+    }
+}

